src/localsocketcontroller.cpp

```diff
@@ -343,9 +353,72 @@ void LocalSocketController::parseCommand(const QByteArray& command) {
   logger.warning() << "Invalid command received:" << command;
 }
 
-void LocalSocketController::write(const QJsonObject& json) {
+void LocalSocketController::write(const QJsonObject& json,
+                                  const QString& expect, int timeout) {
+  QByteArray payload = QJsonDocument(json).toJson(QJsonDocument::Compact);
+  payload.append('\n');
+
+  // If an immediate response to this message is expected, start a timer to
+  // throw an error if that response fails to arrive in a timely manner. This
+  // is used to detect a crash or failure of the daemon.
+  if (!expect.isEmpty()) {
+    QTimer* t = new QTimer(this);
+    connect(t, &QTimer::timeout, this,
+            [&] { this->errorOccurred(QLocalSocket::SocketTimeoutError); });
+
+    t->setProperty("responseType", QVariant(expect));
+    t->setSingleShot(true);
+    t->start(timeout);
+
+    m_expectedResponses.append(t);
+  }
+
   Q_ASSERT(m_socket);
-  m_socket->write(QJsonDocument(json).toJson(QJsonDocument::Compact));
-  m_socket->write("\n");
+  m_socket->write(payload);
   m_socket->flush();
 }
+
+void LocalSocketController::disarmTimeout(const QString& responseType) {
```

Comment:
Multiple requests with the same expected response type are  assumed to receive responsese in the same order. That may be an acceptable assumption. Could a comment be added explaining that?

Reply:
Without modifying the API, we have no way of distinguishing multiple responses of the same type, so this is the best we can do without risking a compatibility issue.

Reply:
Sounds good. Perhaps add a comment with the assumption?